### PR TITLE
feat(textarea): add autoresizing to textarea

### PR DESCRIPTION
**Pull request by dbanksdesign**
#### Description of changes

Adds an `autoResize` prop to the `<TextareaField>` component. 

#### Description of how you validated changes

CleanShot 2024-08-20 at 11 29 35

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] `yarn test` passes and tests are updated/added
- [ ] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by thaddmt — APPROVED** [review]
LGTM!

**Review by thaddmt — DISMISSED** [review]

**Review by reesscot — DISMISSED** [review]

File: packages/react/src/primitives/TextArea/AutoresizeTextarea.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { TextArea } from './TextArea';
+import { useAutoresizeTextArea } from './useAutoresizeTextarea';
+import { useComposeRefsCallback } from '../../hooks/useComposeRefsCallback';
+
+const AutoresizeTextAreaPrimitive: Primitive<TextAreaProps, 'textarea'> = (
+  { value, ...rest },
+  externalRef
+) => {
+  const internalRef = React.useRef<HTMLTextAreaElement>(null);
+  useAutoresizeTextArea(internalRef.current, value as string);
```
Comment (calebpollman) [inline packages/react/src/primitives/TextArea/AutoresizeTextarea.tsx:14]: What happens if `value` is not a `string` (e.g. `undefined`)?

File: packages/react/src/primitives/TextArea/__tests__/AutoresizeTextArea.test.tsx [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        onChange={onChange}
+        onInput={onInput}
+        onPaste={onPaste}
+      />
+    );
+    const textarea = await screen.findByRole('textbox');
+    await act(async () => {
+      await userEvent.type(textarea, 'hello');
+    });
+    // userEvent.paste(textarea, 'there');
```
Comment (calebpollman) [inline packages/react/src/primitives/TextArea/__tests__/AutoresizeTextArea.test.tsx:121]:
```suggestion
```

File: packages/react/src/primitives/TextArea/useAutoresizeTextarea.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  useEffect(() => {
+    const resizeTextArea = () => {
+      if (textAreaRef) {
+        // We need to reset the height momentarily to get the correct scrollHeight for the textarea
+        textAreaRef.style.height = 'auto';
+        const { scrollHeight } = textAreaRef;
+
+        // We then set the height directly, outside of the render loop
+        // Trying to set this with state or a ref will product an incorrect value.
+        textAreaRef.style.height = scrollHeight + 'px';
```
Comment (calebpollman) [inline packages/react/src/primitives/TextArea/useAutoresizeTextarea.ts:17]: Prefer string literal:
```suggestion
        textAreaRef.style.height = `${scrollHeight}px`;
```

File: packages/react/src/primitives/TextArea/AutoresizeTextarea.tsx (hunk 1 above)
Comment (dbanksdesign) [reply packages/react/src/primitives/TextArea/AutoresizeTextarea.tsx:14]: Good point, removing the casting and really the type doesn't matter we only need value to include in the dependency array of an effect

File: packages/react/src/primitives/TextArea/useAutoresizeTextarea.ts (hunk 3 above)
Comment (dbanksdesign) [reply packages/react/src/primitives/TextArea/useAutoresizeTextarea.ts]: donesies

File: packages/react/src/primitives/TextArea/__tests__/AutoresizeTextArea.test.tsx (hunk 2 above)
Comment (dbanksdesign) [reply packages/react/src/primitives/TextArea/__tests__/AutoresizeTextArea.test.tsx]: done

File: packages/react/src/primitives/TextArea/AutoresizeTextarea.tsx (hunk 1 above)
Comment (dbanksdesign) [reply packages/react/src/primitives/TextArea/AutoresizeTextarea.tsx]: done

**Review by thaddmt — APPROVED** [review]

**Review by calebpollman — APPROVED** [review]

**Pull request merged**